Gemini updates (#22)
Ada-Developers-Academy/independent-ai-topics: working-with-ai-apis/ai-apis-in-projects.md

Reviewer: This project already looks like it has some googly stuff in its requirements. Do we need those, since the lesson goes on to have the reader install the genai library?

Author: I think those were left over from the previous installation. I'll remove those from the npc generator!

Reviewer: Similar tidying
```suggestion
    character = validate_model(Character, char_id)
    
    if not character.greetings:
        return {"message": f"No greetings found for {character.name}"}, 201
    
    response = {"character_name": character.name,
                "greetings": []}
    for greeting in character.greetings:
        response["greetings"].append({
            "greeting" : greeting.greeting_text
        })
    
    return response
```

Reviewer: It also feels like we should include the changes to `generate_greetings` greetings for it to work with routes. Here's what I did:
```py
def generate_greetings(character):
    input_message = f"I am writing a fantasy RPG video game. I have an npc named {character.name} who is {character.age} years old. They are a {character.occupation} who has a {character.personality} personality. Please generate a Python style list of 10 stock phrases they might use when the main character talks to them. Please return just the list without a variable name and square brackets."
    response = client.models.generate_content(
        model="gemini-2.5-flash", contents=input_message
    )
    # print(response.text)
    return response.text.splitlines()
```

Reviewer: Also, just over the course of working on this, Gemini returned results in ways that were sometimes incompatible with the code that added the greetings to the character. Maybe we should include a blurb or callout that the results can be unreliably formatted, and that a more orbust implementation would likely specify the desired format more specifically (maybe using prompt shotting) and should have more robust code around handling the result?

Author: There is a callout that mentions this already. I changed it from an info callout to a warn callout however and am fleshing it out to mention some of these points.

</details>
</br>

## Install the `google-genai` Library and Connect Your Secret Key

Before we write our request to the Gemini API within the project, we need to install the `google-genai` package. 

1. In your terminal, run `pip install -q -U google-genai`
2. Run `pip freeze > requirements.txt` to add the `google-genai` package and dependencies to your requirements.
3. In your `.env` file, add the variable `GEMINI_API_KEY` and assign it the value of your secret key: 
    `GEMINI_API_KEY=<Your Secret Key>`

> 4. In the `character_routes.py` file, import the `google-generativeai` package and configure the Gemini API environment using the code below:

Reviewer: ```suggestion
4. In the `character_routes.py` file, import the `google-genai` package and configure the Gemini API environment using the code below:
```

   
    ```python
    from google import genai

>     import os

Reviewer: This is no longer used. It looks like the `genai` client look for `GEMINI_API_KEY` in the environment by default. When I renamed it to something else, the server failed to start up.
```suggestion
```

    client = genai.Client()
    ```
   
   | <div style="min-width:300px;">Part of Code</div> | Description |
   | ------------- | ----------- | 
   | `from google import genai` | Import the `google-genai` package . |

>    | `import os` | Import the `os` package so we can read our `GEMINI_API_KEY` from the `.env` file. |

Reviewer: This can also be removed.
```suggestion
```

   | `client = genai.Client()` | This code creates a Gemini API client with your secret key from the `.env` file. It looks specifically for a key named `GEMINI_API_KEY`. We will use the `genai` import to access the `gemini-2.5-flash` model and make requests to the `/gemini-2.5-flash:generateContent` endpoint. |

## Make the API call

> To make an API call, we'll create a helper function called `generate_greetings` that will take a `Character` in as a parameter. The first thing we will do is construct a prompt for our request body that uses the `Character`'s attributes to describe what we are looking for. use the aliased `google.generativeai` import to get a reference to the AI model we want to use. To do this, we'll call the constructor for the `GenerativeModel` class and pass it the name of the model as an string argument `"gemini-1.5-flash"`.

Reviewer: Left over content.
```suggestion
To make an API call, we'll create a helper function called `generate_greetings` that will take a `Character` in as a parameter. The first thing we will do is construct a prompt for our request body that uses the `Character`'s attributes to describe what we are looking for.
```

}
```

That is a lot of information but now it's up to us to parse through that response and determine what we want to use. Ultimately, we want to just use the `text` within `candidates` > `content` > `parts`. Happily for us, the `response` object returned by `generate_content` has a convenience attribute `.text` that we can use to directly access this `text` key within the JSON. Before we use this response in an endpoint, let's print out the text to see what we get:

```python
print(response.text)
```

This gives our `generate_greetings` helper method the final code of:

```python
def generate_greetings(character):
    input_message = f"I am writing a fantasy RPG video game. I have an npc named {character.name} who is {character.age} years old. They are a {character.occupation} who has a {character.personality} personality. Please generate a Python style list of 10 stock phrases they might use when the main character talks to them. Please return just the list without a variable name and square brackets."
    response = client.models.generate_content(

>         model="gemini-2.5-flash", contents= input_message

Reviewer: ```suggestion
        model="gemini-2.5-flash", contents=input_message
```

This is not the final form our code for this endpoint will take, but if we make a post request to this endpoin in Postman using id 1, the terminal should display an AI generated response to our prompt! Go ahead and make this request now. If your terminal prints out the response, congratulations! You have successfully integrated an AI API into your backend! 

### !callout-warn

## Knowing Your Response
Because our responses are AI generated, we may receive one that is unreliably formatted. Our approach here is pretty basic and has to jump through some hoops to adjust the response we get to fit our needs. A more robust implementation may specify the desired format (e.g. JSON) or use prompt shotting to more accurately describe what is expected. 

### !end-callout

## Using The Response

> Now it's your turn. There are two endpoints left to write. First up is the POST request that creates "Greeting" records for a specific character by using the information returned from the call to the Gemini API. Try using what you know to write the `add_greetings` method. Don't forget to use the `generate_greetings` helper we just created.

Reviewer: Includes a reminder to finish the output of the helper so that the change below isn't out of the blue.
```suggestion
Now it's your turn. There are two endpoints left to write. First up is the POST request that creates "Greeting" records for a specific character by using the information returned from the call to the Gemini API. Try using what you know to write the `add_greetings` method.

Don't forget to use the `generate_greetings` helper we just created, after updating it to return the the output as separate lines.
```

<br />

> <details>

Reviewer: Suggest moving the `generate_greetings` snippet to after the `add_greetings` snippet

Author: Change has been pushed!

>   <summary> Expand to view a modified `generate_greetings` function </summary>

Reviewer: Markdown doesn't work in the summary block
```suggestion
  <summary> Expand to view a modified <code>generate_greetings</code> function that returns its result as separate lines</summary>
```

> ```python
> @bp.post("/<char_id>/generate")
> def add_greetings(char_id):
>     character = validate_model(Character, char_id)
>     greetings = generate_greetings(character)
> 
>     if character.greetings: # Check to see if Greetings have already been added
>         return {"message": f"Greetings already generated for {character.name} "}, 201
>     
>     new_greetings = []
>     
>     for greeting in greetings:
>         new_greeting = Greeting(greeting_text=greeting.strip("\","), # Strip leading and trailing quotes and commas from each greeting
>                                 character = character
>                             )
>         new_greetings.append(new_greeting)
>     
>     db.session.add_all(new_greetings)
>     db.session.commit()
> 
>     return {"message": f"Greetings successfully added to {character.name}"}, 201

Reviewer: Tidied up a bit
```suggestion
    character = validate_model(Character, char_id)

    if character.greetings: # Check to see if Greetings have already been added
        return {"message": f"Greetings already generated for {character.name}"}, 201
    
    greetings = generate_greetings(character)
    new_greetings = []
    
    for greeting in greetings:
        new_greeting = Greeting(greeting_text=greeting.strip("\"',"), # Strip leading and trailing quotes and commas from each greeting
                                character=character
                            )
        new_greetings.append(new_greeting)
    
    db.session.add_all(new_greetings)
    db.session.commit()

    return {"message": f"Greetings successfully added to {character.name}"}, 201

```

Author: 👍
